### PR TITLE
two changes facilitating  sending table/column stats over the wire (M…

### PR DESCRIPTION
**Pull request by peterboncz**
…otherDuck suggested):

- added a de/serialization methods for NodeStatistics
- BaseStatistics may shed the HLL (3KB per col) yet still remember distinct_count

### CODE REVIEW

**Comment by peterboncz** [comment]
Hi DuckDB team. I discorded Mark a few days back with a question about sending table/col stats over the wire, which is the first PR I did for MotherDuck. This is implemented now, but having the small changes in this PR merged in mainline DuckDB would make the MD implementation still better.

File: src/include/duckdb/storage/statistics/node_statistics.hpp [hunk 1]
```diff
@@ -23,6 +24,28 @@ class NodeStatistics {
...
+	static unique_ptr<NodeStatistics> Deserialize(Deserializer &source) {
+		if (!source.Read<bool>()) {
+			return make_unique<NodeStatistics>();
+		}
+		idx_t estimated_cardinality = source.Read<idx_t>();
+		if (!source.Read<bool>()) {
+			return make_unique<NodeStatistics>(estimated_cardinality);
+		}
+		return make_unique<NodeStatistics>(estimated_cardinality, source.Read<idx_t>());
+	}
```
Comment (pdet) [inline src/include/duckdb/storage/statistics/node_statistics.hpp:48]: I think for readibility is usually better to have the `source.Read()` functions returning to a named variable

File: src/include/duckdb/storage/statistics/node_statistics.hpp [hunk 2]
```diff
@@ -23,6 +24,28 @@ class NodeStatistics {
...
+	void Serialize(Serializer &serializer) const { 
+		serializer.Write(has_estimated_cardinality);
+		if (has_estimated_cardinality) {
+			serializer.Write(estimated_cardinality);
+			serializer.Write(has_max_cardinality);
+			if (has_max_cardinality) {
+				serializer.Write(max_cardinality);
+			}
+		} else {
+			assert(!has_max_cardinality);
```
Comment (pdet) [inline src/include/duckdb/storage/statistics/node_statistics.hpp:36]: `D_ASSERT()` instead of `assert`

**Review by pdet — CHANGES_REQUESTED** [review]
Hi Peter, thanks for the contribution! Just have some small comments :-)

File: src/include/duckdb/storage/statistics/node_statistics.hpp (hunk 1 above)
Comment (peterboncz) [reply src/include/duckdb/storage/statistics/node_statistics.hpp]: who am I to disagree on that

File: src/include/duckdb/storage/statistics/node_statistics.hpp (hunk 2 above)
Comment (peterboncz) [reply src/include/duckdb/storage/statistics/node_statistics.hpp]: ah, I saw it did not even survive the CI debug build (but it did on mine, does not compute). So fixed it. And I also ran make format-fix now

**Comment by pdet** [comment]
LGTM, one last thing is that we are currently on a bug fix freeze, which means new features should be merged to the feature branch instead of the master branch. So you should re-tag it to the feature branch. Otherwise it might take a while to be merged.

File: src/include/duckdb/storage/statistics/base_statistics.hpp [hunk 3]
```diff
@@ -39,6 +39,8 @@ class BaseStatistics {
 	unique_ptr<BaseStatistics> validity_stats;
 	//! The approximate count distinct stats of the column (if any)
 	unique_ptr<BaseStatistics> distinct_stats;
+	idx_t distinct_count; // estimate that one may have even if distinct_stats==nullptr
```
Comment (Mytherin) [inline src/include/duckdb/storage/statistics/base_statistics.hpp:42]: nit: Could you use the same commenting style here as in the rest of the code?

File: src/storage/statistics/base_statistics.cpp [hunk 4]
```diff
@@ -72,9 +72,9 @@ void BaseStatistics::Merge(const BaseStatistics &other) {
 idx_t BaseStatistics::GetDistinctCount() {
 	if (distinct_stats) {
 		auto &d_stats = (DistinctStatistics &)*distinct_stats;
-		return d_stats.GetCount();
+		distinct_count = d_stats.GetCount();
 	}
-	return 0;
+	return distinct_count;
```
Comment (Mytherin) [inline src/storage/statistics/base_statistics.cpp:77]: Should the distinct count not also be part of the (de)serialize and copy?

File: src/include/duckdb/storage/statistics/base_statistics.hpp (hunk 3 above)
Comment (Mytherin) [inline src/include/duckdb/storage/statistics/base_statistics.hpp:42]: Could we move the `distinct_count` to the `DistinctStatistics` instead? Then the `DistinctStatistics` could have either a HLL or a count.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

**Comment by Mytherin** [comment]
Actually going to merge this as-is, I'm planning on doing a cleanup/rework of the statistics and I will take these changes along with it.

**Pull request merged**